File: Python/programmers/etc/level2/혼자서_하는_틱택토.py
def solution(board):
    
    def check(board, st):
        for i in range(3):
            if all(board[i][j] == st for j in range(3)):
                return 1
        for j in range(3):
            if all(board[i][j] == st for i in range(3)):
                return 1
        for i in range(3):
            if all(board[i][i] == st for i in range(3)):
                return 1
            elif all(board[i][2-i] == st for i in range(3)):
                return 1
        return 0
    
    o, x = 0, 0
    for i in range(3):
        for j in range(3):
            if board[i][j] == "O":
                o += 1
            elif board[i][j] == "X":
                x += 1
    if o < x or o > x+1:
        return 0
    
    winO, winX = check(board, "O"), check(board, "X")
    
    if winO and winX:
        return 0
    if winO and not winX and o == x:
        return 0
    if not winO and winX and o != x:
        return 0
    else:
        return 1
    

board = ["O.X", ".O.", "..X"]
print(solution(board)) # 1

board = ["OOO", "...", "XXX"]
print(solution(board)) # 0

board = ["...", ".X.", "..."]
print(solution(board)) # 0

board = ["...", "...", "..."]
print(solution(board)) # 1

File: Java/programmers/etc/level2/혼자서하는틱택토.java
package CodeTest.Java.programmers.etc.level2;

public class 혼자서하는틱택토 {
    public int solution(String[] tmp) {
        String[][] board = new String[3][3];
        for (int i=0; i<3; i++) {
            String[] st = tmp[i].split("");
            for (int j=0; j<3; j++) {
                board[i][j] = st[j];
            }
        }

        int o = 0, x = 0;
        for (int i=0; i<3; i++) {
            for (int j=0; j<3; j++) {
                if (board[i][j].equals("O")) o++;
                else if (board[i][j].equals("X")) x++;
            }
        }
        if (o < x || o > x+1) return 0;

        boolean a = check(board, "O"), b = check(board, "X");
        if (a && b) return 0;
        if ((a && !b) && o == x) return 0;
        if ((!a && b) && o != x) return 0;
        return 1;
    }
    
    private boolean check(String[][] board, String st) {
        for (int i=0; i<3; i++) {
            if (board[i][0].equals(st) && board[i][1].equals(st) && board[i][2].equals(st)) return true;
            else if (board[0][i].equals(st) && board[1][i].equals(st) && board[2][i].equals(st)) return true;
        }
        if (board[0][0].equals(st) && board[1][1].equals(st) && board[2][2].equals(st)) return true;
        if (board[0][2].equals(st) && board[1][1].equals(st) && board[2][0].equals(st)) return true;
        return false;
    }
}
